src/common/models/ICUHeadroom.ts

```diff
@@ -0,0 +1,230 @@
+      metricValue: 0.9,
+      overrideInPlace,
+      // rest of this is garbage.
+      totalBeds: 0,
+      covidPatients: 0,
+      covidPatientsIsActual: false,
+      nonCovidPatients: 0,
+      nonCovidPatientsMethod: NonCovidPatientsMethod.ACTUAL,
+    };
+  }
+
+  const covidPatientsResult = calcCovidICUPatientsSeries(
+    dates,
+    actualTimeseries,
+    timeseries,
+    lastUpdated,
+  );
+
+  const nonCovidPatientsResult = calcNonCovidICUPatientsSeries(
+    stateName,
+    dates,
+    actualTimeseries,
+    timeseries,
+    actuals,
+  );
+
+  const totalBeds = actuals.ICUBeds.totalCapacity;
+  const metricSeries = covidPatientsResult.series.map((covidPatients, i) => {
+    const nonCovidPatients = nonCovidPatientsResult.series[i];
+    if (covidPatients === null || nonCovidPatients === null) {
+      return null;
+    } else {
+      return covidPatients / (totalBeds - nonCovidPatients);
+    }
+  });
+
+  const i = indexOfLastValue(metricSeries);
+  if (i === null) {
+    return null;
+  } else {
+    return {
+      metricSeries,
+      metricValue: nonNull(metricSeries[i]),
+      overrideInPlace,
+      totalBeds,
+      covidPatients: nonNull(covidPatientsResult.series[i]),
+      covidPatientsIsActual: covidPatientsResult.isActual,
+      nonCovidPatients: nonNull(nonCovidPatientsResult.series[i]),
+      nonCovidPatientsMethod: nonCovidPatientsResult.method,
+    };
+  }
+}
+
+/**
+ * Calculates the number of covid ICU patients, as a series. Uses actuals if
+ * available.
+ */
+function calcCovidICUPatientsSeries(
+  dates: Array<Date>,
+  actualTimeseries: Array<CANActualsTimeseriesRow | null>,
+  timeseries: Array<CANPredictionTimeseriesRow | null>,
+  lastUpdated: Date,
+): { isActual: boolean; series: Array<number | null> } {
+  const actualCovidPatients = actualTimeseries.map(
+    row => row?.ICUBeds.currentUsageCovid || null,
+  );
+  if (hasRecentData(actualCovidPatients, dates)) {
+    // Return actuals.
+    return { isActual: true, series: actualCovidPatients };
+  } else {
+    // Return Projections.
+    return {
+      isActual: false,
+      series: omitDataAfterDate(
+        timeseries.map(row => row?.ICUBedsInUse || null),
+        dates,
+        lastUpdated,
+      ),
+    };
+  }
+}
+
+/**
+ * Calculates the number of non-covid ICU patients, as a series. Uses actuals
+ * if available.
+ */
+function calcNonCovidICUPatientsSeries(
+  stateName: string,
+  dates: Array<Date>,
+  actualTimeseries: Array<CANActualsTimeseriesRow | null>,
+  timeseries: Array<CANPredictionTimeseriesRow | null>,
+  actuals: _Actuals,
+): { series: Array<number | null>; method: NonCovidPatientsMethod } {
+  const actualTotalPatients = actualTimeseries.map(
+    row => row?.ICUBeds.currentUsageTotal || null,
+  );
+  const actualCovidPatients = actualTimeseries.map(
+    row => row?.ICUBeds.currentUsageCovid || null,
+  );
+  if (hasRecentData(actualTotalPatients, dates)) {
```

Comment:
In theory, the recent actuals might be non-overlapping I guess? That seems pretty unlikely to happen in practice so not worth worrying about.

Reply:
Yeah, that's true.  Something to keep an eye on, I guess.